platform/lib/schema/index.ts

```diff
@@ -436,6 +436,65 @@ export class ObjectSetting<P extends Props> extends Setting<
   }
 }
 
+function isMap<K, V>(o: any): o is Map<K, V> {
+  return Object.prototype.toString.call(o) === '[object Map]';
+}
+
+// We need this helper for the types to be correct
+// (otherwise it assumes an array of A|B instead of a tuple [A,B])
+const toTuple = <A, B>(a: A, b: B): [A, B] => [a, b];
+
+type MapOfOptions<K, V> = SettingOptions<Map<K, V>>;
+
+export class MapOfSetting<K, V> extends Setting<Map<K, V>> {
+  constructor(
+    private readonly keySetting: Setting<K>,
+    private readonly valueSetting: Setting<V>,
+    options: MapOfOptions<K, V> = {}
+  ) {
+    super(options);
+  }
+
+  process(obj: any, context?: string): Map<K, V> {
+    if (isPlainObject(obj)) {
+      const res = Object.keys(obj).map(key => {
+        const validatedKey = this.keySetting.validate(
+          key,
+          toContext(context, key)
+        );
+        const validatedValue = this.valueSetting.validate(
+          obj[key],
+          toContext(context, key)
+        );
+        return toTuple(validatedKey, validatedValue);
+      });
+
+      return new Map(res);
+    }
+
+    if (isMap(obj)) {
+      const res = [...obj.entries()].map(([key, value]) => {
```

Comment:
__nit:__ Hmm, these two pieces look so similar, maybe we can do something like this?
```
let keyValuePairs;

if (isPlainObject(obj)) {
  keyValuePairs = Object.keys(obj).map((key) => [key, obj[key]]);
} else if (isMap(obj)) {
  keyValuePairs = Array.from(obj);
} else {
  throw new SettingError(
    `expected value of type [Map] or [object] but got [${typeDetect(obj)}]`,
    context
  );
}

const res = keyValuePairs.map(([key, value]) => {
  // You see, I don't like local variables if it's easy to read and TS doesn't complain :D
  // But it's personal preference, so up to you 
  return [
    this.keySetting.validate(key, toContext(context, key)),
    this.valueSetting.validate(value, toContext(context, key))
  ] as [K, V];
});

return new Map(res);
```

Although I think additional `map` call is not a big deal in this code (+ we'll get rid of it once we get `Object.entries`), we can go further and iterate lazily if you have performance concerns:
```
let keyValuePairs;

if (isPlainObject(obj)) {
  keyValuePairs = function* iterator() {
    for (const key of Object.keys(obj)) {
      yield [key, obj[key]];
    }
  }();
} else if (isMap(obj)) {
  keyValuePairs = obj;
} else {
  throw new SettingError(
    `expected value of type [Map] or [object] but got [${typeDetect(obj)}]`,
    context
  );
}

const resultMap = new Map();
for (const [key, value] of keyValuePairs) {
  resultMap.set(
    this.keySetting.validate(key, toContext(context, key)),
    this.valueSetting.validate(value, toContext(context, key))
  );
}

return resultMap;
```